OpenAI_API/Chat/Conversation.cs

```diff
@@ -111,34 +112,51 @@ public void AppendMessage(ChatMessage message)
-			req.Messages = _Messages.ToList();
-
-			var res = await _endpoint.CreateChatCompletionAsync(req);
-			MostRecentApiResult = res;
-
-			if (res.Choices.Count > 0)
-			{
-				var newMsg = res.Choices[0].Message;
-				AppendMessage(newMsg);
-				return newMsg.Content;
-			}
-			return null;
-		}
-
-		/// <summary>
-		/// OBSOLETE: GetResponseFromChatbot() has been renamed to <see cref="GetResponseFromChatbotAsync"/> to follow .NET naming guidelines.  This alias will be removed in a future version.
-		/// </summary>
-		/// <returns>The string of the response from the chatbot API</returns>
-		[Obsolete("Conversation.GetResponseFromChatbot() has been renamed to GetResponseFromChatbotAsync to follow .NET naming guidelines.  Please update any references to GetResponseFromChatbotAsync().  This alias will be removed in a future version.", false)]
+        #region Non-streaming
+
+        /// <summary>
+        /// A generic asynchronous helper function that performs a chat request and applies the provided selector function on the resulting <see cref="ChatMessage"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of the data to be returned. It must be a reference type.</typeparam>
+        /// <param name="selector">A function that takes a <see cref="ChatMessage"/> and returns a result of type T.</param>
+        /// <returns>A task that represents the asynchronous operation. The task result contains the selected data from the chat message.</returns>
+        private async Task<T> GetFromChatResultAsync<T>(Func<ChatMessage, T> selector) where T : class
+        {
+            ChatRequest req = new ChatRequest(RequestParameters);
+            req.Messages = _Messages.ToList();
+
+            var res = await _endpoint.CreateChatCompletionAsync(req);
+            MostRecentApiResult = res;
+
+            if (res.Choices.Count > 0)
+            {
+                var message = res.Choices[0].Message;
+                AppendMessage(message);
+                return selector(message);
+            }
+            return null;
+        }
+        /// <summary>
+        /// Calls the API to get a response, which is appended to the current chat's <see cref="Messages"/> as an <see cref="ChatMessageRole.Assistant"/> <see cref="ChatMessage"/>.
+        /// </summary>
+        /// <returns>The string of the response from the chatbot API</returns>
+        public async Task<string> GetResponseFromChatbotAsync()
+        {
+            return await GetFromChatResultAsync(message => message.Content);
+        }
+        /// <summary>
+        /// Calls the API to get a response and retrieves the <see cref="Function_Call"/> from the resulting <see cref="ChatMessage"/>.
+        /// </summary>
+        /// <returns>The Function_Call object from the chat message.</returns>
+        public async Task<Function_Call> GetFunction_CallResponseAsync()
+        {
+            return await GetFromChatResultAsync(message => message.Function_Call);
+        }
```

Comment:
This seems like the wrong way to implement this because it implies you know that the API is going to request a function call based on the input.  The way it's implemented in the python library is to check for an empty `content` return from `GetResponseFromChatbotAsync` and then get `response["choices"][0]["function_call"]` from the last response to check the named Function_Call.

Reply:
Not sure why but I'm not able to find any reference to functions in the openai/openai-python library, if you could link what your going off of I would appreciate it. The cookbook utilizes JSON for examples which we shouldn't be working with as our end result. 

If function_call is not null that implies the existence of a function,. The real question is should we have the user handle the function call or the wrapper handle it. 

If it is null we can just output the response as a string like you said earlier. This is just rambling but we could have the user pass in a list of Func's, then invoke them based on the response. (Which in turn would finally call the summarize). This would circumvent needing to return function_call to the user directly. 

That seems a bit advanced and could be outside the scope but it's an interesting thought.

Alternatively this is all handled just by using 'CreateChatCompletionAsync' and letting the user manually check function_call + content.

Reply:
I would go for the simplest implementation at this point.  I see your point that there really isn't a reference implementation in the python library. It's just relying on the dictionary.  Since there are a dozen ways someone might want to implement handling the presence of function_call in the return - I think the way it works now will suffice.